dirty_cat/_deduplicate.py

```diff
@@ -0,0 +1,175 @@
+
+    Returns
+    -------
+    np.ndarray
+        An n-times-(n-1)/2 matrix of n-gram TfIdf distances between `unique_words`.
+    """
+    enc = CountVectorizer(ngram_range=ngram_range, analyzer=analyzer)
+    encoded = TfidfTransformer().fit_transform(enc.fit_transform(unique_words))
+
+    distance_mat = pdist(encoded.todense(), metric="euclidean")
+    return distance_mat
+
+
+def deduplicate(
+    data: Sequence[str],
+    n_clusters: Optional[int] = None,
+    ngram_range: Tuple[int, int] = (2, 4),
+    analyzer: Literal["word", "char", "char_wb"] = "char_wb",
+    method: Literal[
+        "single", "complete", "average", "centroid", "median", "ward"
+    ] = "average",
+    return_translation_table: bool = False,
+) -> Union[List[str], Tuple[List[str], pd.Series]]:
+    """Deduplicates data by computing the n-gram distance between unique
+    categories in data, performing hierarchical clustering on this distance
+    matrix, and choosing the most frequent element in each cluster as the
+    'correct' spelling. This method works best if the true number of
+    categories is significantly smaller than the number of observed spellings.
+
+    Parameters
+    ----------
+    data : Sequence[str]
+        The data to be deduplicated.
+    n_clusters : Optional[int], optional
+        number of clusters to use for hierarchical clustering, if None use the
+        number of clusters that lead to the lowest silhouette score,
+        by default None
+    ngram_range : Tuple[int, int], optional
+        range to use for computing n-gram distance, by default (2, 4)
+    analyzer : str, optional
+        `CountVectorizer` analyzer for computing n-grams, by default "char_wb"
+    method : str, optional
+        linkage method to use for merging clusters, by default "average"
+    return_translation_table : bool, optional
+        Whether to return a translation table with original words as indices
+        and deduplicated words as values, by default False
+
+    Returns
+    -------
+    Union[List[str], Tuple[List[str], pd.Series]]
+       Either the deduplicated data (if return_translation_table=False) or a
+       tuple of the deduplicated data and a translation table that maps
+       original data to deduplicated data (if return_translation_table=True).
+    """
+    unique_words, counts = np.unique(data, return_counts=True)
+    distance_mat = compute_ngram_distance(
+        unique_words, ngram_range=ngram_range, analyzer=analyzer
+    )
+
+    Z = linkage(distance_mat, method=method, optimal_ordering=True)
+    if n_clusters is None:
+        n_clusters = guess_clusters(Z, distance_mat)
+    clusters = fcluster(Z, n_clusters, criterion="maxclust")
+
+    translation_table = create_spelling_correction(unique_words, counts, clusters)
+    unrolled_corrections = translation_table[data]
+    if return_translation_table:
+        return unrolled_corrections, translation_table
+    else:
+        return unrolled_corrections
```

Comment:
I had the same idea for fuzzy_join, but some users pointed out that the simplest was to return one table, with an additional column containing "non-translated" values. It may be best to do here as well. From experience, the user may then more easily observe results and correct if needed.

Reply:
Not sure I understand, you mean returning translated data as a table with suggested translations (or non-translated values)?
Like `translation_table`, just for all data instead of the unique examples (`translation_table` contains unique strings/categories + suggested translations, or the original if no translation was suggested)?

Reply:
I actually meant returning a merged table `unrolled_corrections` + `translation_table` if `return_translation_table` is True.

Reply:
Ok actually, I think the best is that this function only returns `unrolled_corrections table`. 
The main problem is that the output format changes depending on parameters value, which is usually not recommended.
Translation tables can be easily extracted then by dropping duplicates in the dirty data column..

Reply:
I agree, it should always return the same output type.
Okay, will return the table you described above with translated/corrected and uncorrected values (lmk if I misunderstood).